File: quick_sort/quick_sort1.py
def partition(A, start, end):
    # boundary between elements less than and greater than pivot
    i = start + 1

    # boundary between partitioned and unpartitioned array
    j = start + 1

    while (j < len(A)):
        if (A[j] < A[start]):
            A[j], A[i] = A[i], A[j]
            i += 1
        j += 1

    A[start], A[i - 1] = A[i - 1], A[start]

    return A, i - 1

def quick_sort2(A, start, end):
    if(start>=end):
        if(start<len(A) and start>=0):
            return [A[start]]
    else:
        A, partition_index = partition(A, start, end)
        quick_sort2(A, start, partition_index-1)
        quick_sort2(A, partition_index+1, end)

        return A

def quick_sort(A):
    return quick_sort2(A, 0, len(A)-1)

# A=[8, 10, 7, 1, 12, 3, 2, 23, 40, 21, 14, 52, 47]
# print(quick_sort(A))

File: quick_sort/quickSort1.c
#include <stdio.h>
#include <stdlib.h>

int* A=NULL;
int n;

int partition(int start, int end, int pivot){
    int* newA=(int *)calloc(n, sizeof(int));
    int k=start;
    int pivotPos;
    int i;

    for(i=0;i<start;i++){
        *(newA+i)=*(A+i);
    }

    for(i=end;i<n;i++){
        *(newA+i)=*(A+i);
    }

    for(i=start;i<end+1;i++){
        if(*(A+i)<pivot){
            *(newA+k)=*(A+i);
            k++;
        }
    }
    pivotPos=k;
    *(newA+k)=pivot;
    k++;

    i=0;
    while(k<end+1){
        if(*(A+i)>pivot){
            *(newA+k)=*(A+i);
            k++;
        }
        i++;
    }
    A=newA;
    return pivotPos;
}

void quickSort(int start, int end){
    int pivot;
    int pivotPos;
    if(end==start){
        return;
    }else{
        pivot=*(A+start);
        pivotPos=partition(start, end, pivot);
        quickSort(start, pivotPos);
        quickSort(pivotPos+1, end);
    }
}

int main(){
    int i;

    scanf("%d", &n);
    A=(int *)calloc(n, sizeof(int));
    for(i=0;i<n;i++){
        scanf("%d", (A+i));
    }
    quickSort(0, n-1);
    // partition(1, 3, *(A+2));
    for(i=0;i<n;i++){
        printf("%d ", *(A+i));
    }
    return 0;
}